### PR TITLE
Feature/trucks

### PR DESCRIPTION
**Pull request by asariev**

### CODE REVIEW

File: src/main/java/io/onroad/eventconsumer/repository/VehicleRepo.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.data.repository.CrudRepository;
+import org.springframework.stereotype.Repository;
+
+import java.util.Optional;
+
+@Repository
+public interface VehicleRepo extends CrudRepository<Vehicle, Long> {
+
+    @Override
+    <S extends Vehicle> S save(S entity);
```
Comment (onroadeld) [inline src/main/java/io/onroad/eventconsumer/repository/VehicleRepo.java:13]: Do we need it here?

File: src/main/java/io/onroad/eventconsumer/repository/VehicleRepo.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package io.onroad.eventconsumer.repository;
+
+import io.onroad.eventconsumer.domain.model.Vehicle;
+import org.springframework.data.repository.CrudRepository;
+import org.springframework.stereotype.Repository;
+
+import java.util.Optional;
+
+@Repository
+public interface VehicleRepo extends CrudRepository<Vehicle, Long> {
```
Comment (onroadeld) [inline src/main/java/io/onroad/eventconsumer/repository/VehicleRepo.java:10]: Why CrudRepository and not JPARepository?

File: src/main/java/io/onroad/eventconsumer/service/VehicleService.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+public interface VehicleService {
+
+    VehicleView create (User user, CreateVehicleRequest createVehicleRequest);
+
+    VehicleView update (Long id, CreateVehicleRequest createVehicleRequest);
+
+    VehicleView delete (Long id);
+
+    VehicleView getVehicle (Long id);
```
Comment (onroadeld) [inline src/main/java/io/onroad/eventconsumer/service/VehicleService.java:17]: Why we need an interface?

File: src/main/java/io/onroad/eventconsumer/service/implementation/CompanyServiceImpl.java [hunk 4]
```diff
@@ -18,19 +20,20 @@
...
     @Override
     public CompanyView update(long id, CreateCompanyRequest companyRequest) {
-        Optional<Company> company = companyRepo.findById(id);
-        if(company.isEmpty()){
-            return null;
-        }
-        return companyViewMapper.viewCompany(companyRepo.save(company.get()));
+        companyRequestValidator.validate(id, companyRequest);
+        Company company = companyRepo.findById(id).get();
+        companyEditMapper.udpate(company, companyRequest);
```
Comment (onroadeld) [inline src/main/java/io/onroad/eventconsumer/service/implementation/CompanyServiceImpl.java:35]: I think here is some typo - udpate

File: src/main/java/io/onroad/eventconsumer/service/implementation/CompanyServiceImpl.java [hunk 5]
```diff
@@ -39,13 +42,13 @@ public CompanyView delete(long id) {
         if(company.isEmpty()){
             return null;
         }
-        CompanyView companyView = companyViewMapper.viewCompany(company.get());
-        companyRepo.deleteById(company.get().getId());
-        return companyView;
+        company.get().setEnabled(false);
+        companyRepo.save(company.get());
+        return companyViewMapper.viewCompany(company.get());
```
Comment (onroadeld) [inline src/main/java/io/onroad/eventconsumer/service/implementation/CompanyServiceImpl.java:47]: Deleting means deactivating? Why not simply delete instead? Who will be able to delete the company? And why return the deleted company back in the response?

**Review by onroadeld — DISMISSED** [review]
Left some comments

File: src/main/java/io/onroad/eventconsumer/api/VehicleApi.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+import io.onroad.eventconsumer.service.VehicleService;
+import io.swagger.v3.oas.annotations.tags.Tag;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.core.annotation.AuthenticationPrincipal;
+import org.springframework.web.bind.annotation.*;
+
+import javax.annotation.security.RolesAllowed;
+import javax.validation.Valid;
+
+@Tag(name = "vehicle")
```
[the remark this answers is not in the document]
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/api/VehicleApi.java:17]: Dos, you mean, you want me to update `api.yml` file too?

File: src/main/java/io/onroad/eventconsumer/domain/dto/CompanyView.java [hunk 7]
```diff
@@ -1,11 +1,20 @@
...
 
 @Data
 public class CompanyView {
 
     private String id;
-    private String email;
     private String name;
+    private String mcNumber;
+    private String usDotNumber;
+    private String address;
```
[the remark this answers is not in the document]
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/domain/dto/CompanyView.java]: updated

File: src/main/java/io/onroad/eventconsumer/domain/dto/CreateEldRequest.java [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package io.onroad.eventconsumer.domain.dto;
+
+import lombok.Data;
+
+@Data
+public class CreateEldRequest {
+
+    private String registration_id;
```
[the remark this answers is not in the document]
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/domain/dto/CreateEldRequest.java]: updated

File: src/main/java/io/onroad/eventconsumer/domain/mapper/VehicleEditMapper.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Autowired
+    private CompanyRepo companyRepo;
+
+    public Vehicle create (CreateVehicleRequest vehicleRequest){
+        if(vehicleRequest == null){
+            return null;
+        }
+        Vehicle vehicle = new Vehicle();
+        Optional<Company> company = companyRepo.findById(vehicleRequest.getCompanyId());
```
[the remark this answers is not in the document]
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/domain/mapper/VehicleEditMapper.java:23]: Validated in CreateVahicleValitodator class

File: src/main/java/io/onroad/eventconsumer/domain/mapper/VehicleViewMapper.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        }
+
+        VehicleView vehicleView = new VehicleView();
+        vehicleView.setId(vehicle.getId());
+        vehicleView.setVIN(vehicle.getVIN());
+        vehicleView.setLicensePlate(vehicle.getLicensePlate());
+        vehicleView.setLicensePlateState(vehicle.getLicensePlateState());
+        vehicleView.setMake(vehicle.getMake());
+        vehicleView.setModel(vehicle.getModel());
+        vehicleView.setCompanyId(vehicle.getCompany().getId());
```
[the remark this answers is not in the document]
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/domain/mapper/VehicleViewMapper.java:23]: I believe I used for `JUnit Tests` in order to delete, get ...

File: src/main/java/io/onroad/eventconsumer/repository/VehicleRepo.java (hunk 1 above)
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/repository/VehicleRepo.java:13]: I believe I used for saving the Vehicle

File: src/main/java/io/onroad/eventconsumer/service/VehicleService.java (hunk 3 above)
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/service/VehicleService.java:17]: Dos backend bolup jurgondo ushundai structure menen tuzup uironup kalgan ekem, `@Tag` koldonsom dele bolso kerek

File: src/main/java/io/onroad/eventconsumer/service/implementation/CompanyServiceImpl.java (hunk 4 above)
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/service/implementation/CompanyServiceImpl.java]: updated

File: src/main/java/io/onroad/eventconsumer/service/implementation/CompanyServiceImpl.java (hunk 5 above)
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/service/implementation/CompanyServiceImpl.java:47]: Only `SUPER_ADMIN` can CRUD company, For Admin, we would like to keep all the companies that we were and are working

File: src/main/java/io/onroad/eventconsumer/domain/model/Eld.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.NoArgsConstructor;
+
+import javax.persistence.*;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Entity
+@Table(name = "ELD_device")
+public class Eld {
```
[the remark this answers is not in the document]
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/domain/model/Eld.java]: udpated

File: src/main/java/io/onroad/eventconsumer/domain/model/User.java [hunk 12]
```diff
@@ -76,6 +76,9 @@
   @JoinColumn(name = "company_id", nullable = false, referencedColumnName = "id")
   private Company company;
 
+  @ManyToOne(fetch = FetchType.LAZY)
+  @JoinColumn(name = "vehicle_id", nullable = false, referencedColumnName = "id")
+  private Vehicle vehicle;
```
[the remark this answers is not in the document]
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/domain/model/User.java:80]: yes

File: src/main/java/io/onroad/eventconsumer/domain/model/Vehicle.java [hunk 13]
```diff
@@ -23,7 +23,27 @@
...
+    private String licensePlate;
+    @Enumerated(EnumType.STRING)
+    private State licensePlateState;
+    @NotNull
+    private String VIN;
+    private String make;
+    private Long year;
+    private String model;
+
+    @OneToMany(mappedBy = "vehicle", cascade = CascadeType.ALL, orphanRemoval = true)
```
[the remark this answers is not in the document]
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/domain/model/Vehicle.java:36]: I think, it should be safe for `@OneToMany` to user Cascade ALL, when we delete vehicle, users should be detached. Any suggestion

File: src/main/java/io/onroad/eventconsumer/repository/CompanyRepo.java [hunk 14]
```diff
@@ -18,4 +20,15 @@
...
+
+    default Company getById(long id) {
+        Optional<Company> companyOptional = findById(id);
+        if (companyOptional.isEmpty()) {
+            throw new NotFoundException(Company.class, id);
+        }
+        if (!companyOptional.get().isEnabled()) {
+            throw new NotFoundException(Company.class, id);
+        }
+        return companyOptional.get();
```
[the remark this answers is not in the document]
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/repository/CompanyRepo.java]: udpated

File: src/main/java/io/onroad/eventconsumer/repository/VehicleRepo.java (hunk 2 above)
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/repository/VehicleRepo.java:10]: I can use JPARepository, if we need more functions, like pagination

File: src/main/java/io/onroad/eventconsumer/api/VehicleApi.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (asariev) [reply src/main/java/io/onroad/eventconsumer/api/VehicleApi.java:17]: updated

File: src/main/java/io/onroad/eventconsumer/domain/model/Vehicle.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (nursultanturdaliev) [reply src/main/java/io/onroad/eventconsumer/domain/model/Vehicle.java:36]: This means users also get deleted.

File: src/main/java/io/onroad/eventconsumer/domain/model/Vehicle.java [hunk 15]
```diff
@@ -23,7 +23,27 @@
...
+    private List<User> user;
+
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "electronicLoggingDevice_id", referencedColumnName = "id")
+    private ElectronicLoggingDevice electronicLoggingDevice;
+
+    @ManyToOne( fetch = FetchType.LAZY)
+    @JoinColumn(name = "company_id", nullable = false, referencedColumnName = "id")
+    private Company company;
+
```
Comment (nursultanturdaliev) [inline src/main/java/io/onroad/eventconsumer/domain/model/Vehicle.java:46]: So the vehicle belongs to a company and a user? What happens for co-drivers?

**Review by nursultanturdaliev — APPROVED** [review]

**Pull request merged**